### PR TITLE
#247 Hash API Methods Implementation

### PR DESCRIPTION
**Pull request by Maxwell-Alexius**

### CODE REVIEW

File: vm/hash.go [hunk 1]
```diff
@@ -231,6 +635,59 @@ func (h *HashObject) length() int {
...
+func (h *HashObject) sortedKeys() []string {
+	var arr []string
+	for k := range h.Pairs {
+		arr = append(arr, k)
+	}
+	sort.Strings(arr)
+	return arr
+}
+
+func (h *HashObject) isEqualTo(compare *HashObject) bool {
```
Comment (Maxwell-Alexius) [inline vm/hash.go]: Note: I think this has a better implementation

Comment (st0012) [inline vm/hash.go]: Can we use [reflect.DeepEqual](https://golang.org/pkg/reflect/#DeepEqual) in this case?

Comment (Maxwell-Alexius) [inline vm/hash.go]: Let me try it ~ I'll also look for the information about reflect in Go

File: vm/hash.go [hunk 2]
```diff
@@ -133,24 +134,267 @@ func builtinHashInstanceMethods() []*BuiltInMethodObject {
...
+			// @return [Array]
+			Name: "each_key",
+			Fn: func(receiver Object) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *callFrame) Object {
+					if len(args) != 0 {
+						return initErrorObject(ArgumentErrorClass, "Expect 0 argument. got=%d", len(args))
+					}
+
+					if blockFrame == nil {
+						t.returnError("Can't yield without a block")
```
Comment (Maxwell-Alexius) [inline vm/hash.go]: @st0012 Would you suggest me to change it to something like this?

```
return initErrorObject(InternalError, "Can't yield without a block")
```

Comment (st0012) [inline vm/hash.go]: https://github.com/goby-lang/goby/commit/9a428b14be809d5c7a4e850a4a1ce04b566996f0
We should improve this method to make it receive error type argument instead

**Comment by hachi8833** [comment]
Checked out `feature/issue-247`. LGTM.

**Comment by Maxwell-Alexius** [comment]
I'll rebase the branch first

File: vm/thread.go [hunk 3]
```diff
@@ -174,7 +174,8 @@ func (t *thread) evalMethodObject(receiver Object, method *MethodObject, receive
 
 // TODO: Use this method to replace unnecessary panics
 func (t *thread) returnError(msg string) {
-	err := &Error{Message: msg}
+	err := initErrorObject(InternalErrorClass, msg)
+	//err := &Error{Message: msg}
```
Comment (Maxwell-Alexius) [inline vm/thread.go]: Changed the `err` to the `initErrorObject` format

File: vm/hash.go [hunk 4]
```diff
@@ -200,8 +689,12 @@ func (h *HashObject) toString() string {
 	var out bytes.Buffer
 	var pairs []string
 
-	for key, value := range h.Pairs {
-		pairs = append(pairs, fmt.Sprintf("%s: %s", key, value.toString()))
+	for _, key := range h.sortedKeys() {
+		if _, isString := h.Pairs[key].(*StringObject); isString {
```
Comment (st0012) [inline vm/hash.go]: I don't like that we need to check this for every pair. But I don't have better idea to improve this yet.

File: vm/hash_test.go [hunk 5]
```diff
@@ -3,42 +3,650 @@ package vm
...
+		evaluated := vm.testEval(t, tt.input)
+		testArrayObject(t, i, evaluated, tt.expected)
+		vm.checkCFP(t, i, 0)
+	}
+}
+
+func TestHashEachKeyMethodFail(t *testing.T) {
+	vm := initTestVM()
+
+	testArgumentError := `
```
Comment (st0012) [inline vm/hash_test.go]: I think we better not writing test in this way. We should still uses table structure for this like
```go
tests := []struct{
    input string
    expectedErr string
    expectedCfp int
} { ..... }
```

Comment (Maxwell-Alexius) [inline vm/hash_test.go]: Okay, I will change this

File: vm/hash.go (hunk 4 above)
Comment (Maxwell-Alexius) [inline vm/hash.go]: I think I'll mark this as a TODO

**Pull request merged**